lib/launcher/main.c

```diff
@@ -0,0 +1,353 @@
+#include "cuda.h"
+#include <Python.h>
+
+union ArgUnion {
+  PyObject *O;
+  float f;
+  double d;
+  long l;
+  uint32_t I;
+  int32_t i;
+  uint64_t K;
+  int64_t L;
+  CUdeviceptr CUdptr;
+};
+
+static inline void gpuAssert(CUresult code, const char *file, int line) {
+  if (code != CUDA_SUCCESS) {
+    const char *prefix = "Triton Error [CUDA]: ";
+    const char *str;
+    cuGetErrorString(code, &str);
+    char err[1024] = {0};
+    strcat(err, prefix);
+    strcat(err, str);
+    PyErr_SetString(PyExc_RuntimeError, err);
+  }
+}
+#define CUDA_CHECK(ans)                                                        \
+  { gpuAssert((ans), __FILE__, __LINE__); }
+
+void my_launch(int gridX, int gridY, int gridZ, int num_warps,
+               int shared_memory, CUstream stream, CUfunction function,
+               void **params) {
+  if (gridX * gridY * gridZ > 0) {
+    CUDA_CHECK(cuLaunchKernel(function, gridX, gridY, gridZ, 32 * num_warps, 1,
+                              1, shared_memory, stream, params, 0));
+  }
+}
+
+static inline CUdeviceptr getPointer(PyObject *obj, int idx) {
+  if (PyLong_Check(obj)) {
+    return (CUdeviceptr)PyLong_AsUnsignedLongLong(obj);
+  }
+  if (obj == Py_None) {
+    return (CUdeviceptr)0;
+  }
+  PyObject *ptr = PyObject_GetAttrString(obj, "data_ptr");
+  if (ptr) {
+    PyObject *empty_tuple = PyTuple_New(0);
+    PyObject *ret = PyObject_Call(ptr, empty_tuple, NULL);
+    Py_DECREF(empty_tuple);
+    Py_DECREF(ptr);
+    if (!PyLong_Check(ret)) {
+      PyErr_SetString(
+          PyExc_TypeError,
+          "data_ptr method of Pointer object must return 64-bit int");
+    }
+    return (CUdeviceptr)PyLong_AsUnsignedLongLong(ret);
+  }
+  PyErr_SetString(
+      PyExc_TypeError,
+      "Pointer argument must be either uint64 or have data_ptr method");
+  return (CUdeviceptr)0;
+}
+
+static PyObject *launch(PyObject *self, PyObject *args) {
+  char *format = (char *)PyUnicode_AsUTF8(PyTuple_GetItem(args, 0));
+  char *is_const = (char *)PyUnicode_AsUTF8(PyTuple_GetItem(args, 1));
+  int gridX, gridY, gridZ;
+  uint64_t _stream;
+  uint64_t _function;
+  int num_warps;
+  int shared_memory;
+  PyObject *launch_enter_hook = NULL;
+  PyObject *launch_exit_hook = NULL;
+  PyObject *compiled_kernel = NULL;
+  PyObject *hook_ret = NULL;
+
+  int ArgC = strlen(format) - strlen("ssiiiiiKKOOO");
+  union ArgUnion MyArgs[100];
+
+  if (0)
+    printf("shal1t7 ArgC = %d\n", ArgC);
+
+  if (0)
+    printf("shal1t7 format = %s\n", format);
+
+  if (0)
+    printf("shal1t7  is_const = %s\n", is_const);
+
+  char *arg0;
+  char *arg1;
+
+  if (ArgC == 1 &&
```

Comment:
Enumerating over number of args doesn't seem efficient in terms of execution. 
And the code is also longer than our previous approach.

Reply:
(the compiler will actually optimize it, but in any case at the source level it's easy to turn it into a 'switch', so the only overhead will be `strlen` + jump, but this is negligible in comparison with calls to python / cuda)
(unfortunately, some upper bound (we can make it very large) for the number arguments seems to be unavoidable (but I'm open to suggestions),
because of the way python arguments are parsed/passed around)

Reply:
> (the compiler will actually optimize it, but in any case at the source level it's easy to turn it into a 'switch', so the only overhead will be strlen + jump, but this is negligible in comparison with calls to python / cuda)

Can you perform some evaluation to compare the overhead of the static launcher against our existing solution? Only the C++ side timing is enough. 

> (unfortunately, some upper bound (we can make it very large) for the number of arguments seems to be unavoidable (but I'm open to suggestions),

Yeah, I'm not sure what would be the optimal solution. We can make the upper bound something like 256, which could be safe in most cases. 

I do think this PR has value to reduce the CI running time since we don't have to recompile launchers. Do you have any suggestions @ptillet ?